### PR TITLE
Network simulation for swarm tests

### PR DESCRIPTION
**Pull request by janos**
This is a proof of concept.

### CODE REVIEW

File: swarm/netsim/connect.go [hunk 1]
```diff
@@ -0,0 +1,159 @@
...
+	for i := 0; i < l-1; i++ {
+		err = s.connect(ids[i], ids[i+1])
+		if err != nil {
+			return err
+		}
+	}
+	return s.connect(ids[l-1], ids[0])
+}
+
+// ConnectNodesStar connects all nodes in a start topology
```
Comment (zelig) [inline swarm/netsim/connect.go]: start -> star

File: swarm/netsim/kademlia.go [hunk 2]
```diff
@@ -0,0 +1,89 @@
...
+	"github.com/ethereum/go-ethereum/log"
+	"github.com/ethereum/go-ethereum/swarm/network"
+)
+
+// BucketKeyKademlia is the key to be used for storing the kademlia
+// instance for particuar node, usually inside the ServiceFunc function.
+var BucketKeyKademlia BucketKey = "kademlia"
+
+// WaitKademlia is blocking until the health of all kademlias is true.
+func (s *Simulation) WaitKademlia(ctx context.Context, kadMinProxSize int) (err error) {
```
Comment (zelig) [inline swarm/netsim/kademlia.go]: this should be callled `WaitTillHealthy`

File: swarm/netsim/simulation.go [hunk 3]
```diff
@@ -0,0 +1,262 @@
...
+
+	httpSrv *http.Server        //attach a HTTP server via SimulationOptions
+	handler *simulations.Server //HTTP handler for the server
+	runC    chan struct{}       //channel where frontend signals it is ready
+}
+
+//SimulationOptions defines options for the Simulation
+//Currently only defines options to allow to attach a
+//HTTP server to the simulation (i.e. for sim frontend visualization)
+type SimulationOptions struct {
```
Comment (zelig) [inline swarm/netsim/simulation.go]: first, here simulation is suddenly used in a differet sense.
i would consider the `with pattern here` 

`simulation.New(services).WithServer(port)`

File: swarm/netsim/simulation.go [hunk 4]
```diff
@@ -0,0 +1,262 @@
...
+// The first argument provides ServiceContext from the adapters package
+// giving for example the access to NodeID. Second argument is the sync.Map
+// where all "global" state related to the service should be kept.
+// All cleanups needed for constructed service and any other constructed
+// objects should ne provided in a single returned cleanup function.
+type ServiceFunc func(ctx *adapters.ServiceContext, bucket *sync.Map) (s node.Service, cleanup func(), err error)
+
+// NewSimulation creates a new Simulation instance with new
+// simulations.Network initialized with provided services.
+func NewSimulation(services map[string]ServiceFunc, opts *SimulationOptions) (s *Simulation) {
```
Comment (zelig) [inline swarm/netsim/simulation.go]: just `New` if in `swarm/network/simulation`

File: swarm/netsim/simulation.go [hunk 5]
```diff
@@ -0,0 +1,262 @@
...
+
+	s.Net = simulations.NewNetwork(
+		adapters.NewSimAdapter(adapterServices),
+		&simulations.NetworkConfig{ID: "0"},
+	)
+
+	if opts == nil {
+		opts = new(SimulationOptions)
+	}
+	if opts.WithHTTP {
```
Comment (zelig) [inline swarm/netsim/simulation.go]: consider putting http server related stuff in a separate file `server.go` or `http.go`

File: swarm/netsim/simulation.go [hunk 6]
```diff
@@ -0,0 +1,262 @@
...
+	return nil
+}
+
+//register additional HTTP routes
+func (s *Simulation) addSimulationRoutes() {
+	s.handler.POST("/runsim", s.RunSimulation)
+}
+
+// StartNetwork starts all nodes in the network
+func (s *Simulation) RunSimulation(w http.ResponseWriter, req *http.Request) {
```
Comment (zelig) [inline swarm/netsim/simulation.go]: avoid using 'simulation' in this meaning

**Review by zelig — APPROVED** [review]
beautiful PR

File: swarm/netsim/simulation.go [hunk 7]
```diff
@@ -0,0 +1,262 @@
...
+}
+
+// Run calls the RunFunc function while taking care of
+// cancelation provided through the Context.
+func (s *Simulation) Run(ctx context.Context, f RunFunc) (r Result) {
+	//if the option is set to run a HTTP server with the simulation,
+	//init the server and start it
+	start := time.Now()
+	if s.httpSrv != nil {
+		err := s.startHTTPServer(ctx)
```
Comment (janos) [inline swarm/netsim/simulation.go]: @holisticode we should not start http server in Run method, as we can call Run multiple time and all but first call will not error. Maybe to add WithServer method that will start the servers instead, as @zelig suggested in https://github.com/ethersphere/go-ethereum/pull/769/files#r199888164 comment?

File: swarm/netsim/simulation.go (hunk 5 above)
Comment (janos) [reply swarm/netsim/simulation.go]: @holisticode http server related things are now in separate file and test also, but I did not address other @zelig comments regarding http server

Comment (holisticode) [reply swarm/netsim/simulation.go]: By other comments you mean the "WithServer" pattern?

Comment (janos) [reply swarm/netsim/simulation.go]: This one too https://github.com/ethersphere/go-ethereum/pull/769#discussion_r199889140

**Comment by janos** [comment]
@zelig, thank you for review. the code is now updated with more functions, tests and with @holisticode's work. Could you review it again? And if I remember correctly, you suggested to call this package `swarm/simulation`, should `swarm/netsim` be renamed?

File: swarm/netsim/events_test.go [hunk 8]
```diff
@@ -0,0 +1,100 @@
...
+package netsim
+
+import (
+	"context"
+	"sync"
+	"testing"
+	"time"
+)
+
+func TestPeerEvents(t *testing.T) {
```
Comment (holisticode) [inline swarm/netsim/events_test.go]: Please add a short description for what the test does if the test does a number of steps, e.g.
- Create simulation
- Add 2 nodes
- Register for PeerEvents
- Connect peers in chain
- Check amount of expected events  received, should match

Not needed for all tests of course (there are tests which really it is easy to understand, like `ExampleSimulation_WaitTillHealthy`

File: swarm/netsim/example_test.go [hunk 9]
```diff
@@ -0,0 +1,140 @@
...
+	"github.com/ethereum/go-ethereum/node"
+	"github.com/ethereum/go-ethereum/p2p"
+	"github.com/ethereum/go-ethereum/p2p/simulations/adapters"
+	"github.com/ethereum/go-ethereum/swarm/netsim"
+	"github.com/ethereum/go-ethereum/swarm/network"
+)
+
+// Every node can have a Kademlia associated using the node bucket under
+// BucketKeyKademlia key. This allows to use WaitTillHealthy to block until
+// all nodes have the their Kadmlias healthy.
```
Comment (holisticode) [inline swarm/netsim/example_test.go]: Please add a short description of how to run this tests, as they are not "normal" tests (the same way you explained to me in the chat)

**Review by holisticode — APPROVED** [review]
Minor comments. The package is still `/netsim` though, did we reach a conclusion for the name?

File: swarm/netsim/bucket.go [hunk 10]
```diff
@@ -0,0 +1,81 @@
...
+// along with the go-ethereum library. If not, see <http://www.gnu.org/licenses/>.
+
+package netsim
+
+import (
+	"github.com/ethereum/go-ethereum/p2p/discover"
+)
+
+// BucketKey is the type that should be used for keys in simulation buckets.
+type BucketKey string
```
[the remark this answers is not in the document]
Comment (janos) [reply swarm/netsim/bucket.go]: No, it is required for BucketKeyKademlia, but strings can be used for keys in specific tests.

File: swarm/netsim/example_test.go (hunk 9 above)
Comment (janos) [reply swarm/netsim/example_test.go]: Actually, this text can not be run, as it does not fulfill requirements to be executed (no // Output: comment). There is another test fo testing kademlia TestWaitTillHealthy. Using examples as test is well described here https://blog.golang.org/examples. But in this case, it is better to test it properly in a dedicated test, and have examples for documentation.

**Comment by janos** [comment]
@holisticode One suggestion is to call it `simulation` https://github.com/ethersphere/go-ethereum/pull/769#issuecomment-403857312.

**Comment by janos** [comment]
As we discussed and decided, the package has been renamed to `swarm/network/simulation`.

**Pull request merged**